src/server/renderer.js

```diff
@@ -32,28 +33,54 @@ export default (projectConfig, projectToolsConfig) => {
     const store = createStore(middleware, history);
 
     if (__DISABLE_SSR__) {
-      const content = html(config, tools.assets(), store, res._headers);
-      res.status(200).send(content);
+      const content = html(config, tools.assets(), store, headers);
+      send(200, content);
     } else {
-      match({ history, routes: getRoutes(store), location: req.originalUrl }, (error, redirectLocation, renderProps) => {
+      match({ history, routes: getRoutes(store), location: originalUrl }, (error, redirectLocation, renderProps) => {
         if (redirectLocation) {
-          res.redirect(redirectLocation.pathname + redirectLocation.search);
+          redirect(redirectLocation.pathname + redirectLocation.search);
         } else if (error) {
           console.error('ROUTER ERROR:', pretty.render(error));
-          res.status(500);
+          send(500);
         } else if (renderProps) {
           rootComponent.createForServer(store, renderProps)
             .then(({ root }) => {
-              const content = html(config, tools.assets(), store, res._headers, root);
-              res.status(200).send(content);
+              const content = html(config, tools.assets(), store, headers, root);
+              send(200, content);
             })
             .catch((err) => {
-              res.status(500).send(err);
+              console.log('ERROR GENERATING ROOT COMPONENT', err, err.stack);
+              send(500, err);
             });
         } else {
-          res.status(404).send('Not found');
+          send(400, 'Not found');
         }
       });
     }
   };
+
+  function *koaMiddleware() {
+    dynamicMiddleware(this.request.originalUrl,
+      this.request.headers,
+      (status, body) => {
+        this.status = status;
+        this.body = body;
+      },
+      (url) => this.response.redirect(url));
+  }
+
+  switch (config.server.webFramework) {
+    case 'koa': {
+      return koaMiddleware;
+    }
+    default:
+    case 'express': {
+      return (req, res) => {
+        dynamicMiddleware(req.originalUrl,
+          req._headers,
```

Comment:
@bartolkaruza This should be `res._headers` ... the JWT example relies on headers and was broken as is.